dkg/dkg_test.go

```diff
@@ -0,0 +1,199 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package dkg_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/coinbase/kryptology/pkg/signatures/bls/bls_sig"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/cluster"
+	"github.com/obolnetwork/charon/cmd"
+	"github.com/obolnetwork/charon/dkg"
+	"github.com/obolnetwork/charon/p2p"
+	"github.com/obolnetwork/charon/tbls"
+	"github.com/obolnetwork/charon/testutil"
+	"github.com/obolnetwork/charon/testutil/keystore"
+)
+
+func TestDKG(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const (
+		nodes = 3
+		vals  = 2
+	)
+
+	// Start bootnode
+	bootnode, errChan := startBootnode(ctx, t)
+
+	// Setup
+	lock, keys, _ := cluster.NewForT(t, vals, 3, 3, 0)
+
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+
+	conf := dkg.Config{
+		DataDir: dir,
+		P2P: p2p.Config{
+			UDPBootnodes: []string{bootnode},
+		},
+		Log:     log.DefaultConfig(),
+		TestDef: &lock.Definition,
+	}
+
+	// Run dkg for each node
+	var eg errgroup.Group
+	for i := 0; i < nodes; i++ {
+		conf := conf
+		conf.DataDir = path.Join(dir, fmt.Sprintf("node%d", i))
+		conf.P2P.TCPAddrs = []string{testutil.AvailableAddr(t).String()}
+		conf.P2P.UDPAddr = testutil.AvailableAddr(t).String()
+
+		require.NoError(t, os.MkdirAll(conf.DataDir, 0o755))
+		err := crypto.SaveECDSA(p2p.KeyPath(conf.DataDir), keys[i])
+		require.NoError(t, err)
+
+		eg.Go(func() error {
+			return dkg.Run(ctx, conf)
+		})
+		if i == 0 {
+			// Allow node0 some time to startup, this just mitigates startup races and backoffs but isn't required.
+			time.Sleep(time.Millisecond * 100)
+		}
+	}
+
+	// Wait until complete
+
+	runChan := make(chan error, 1)
+	go func() {
+		runChan <- eg.Wait()
+	}()
+
+	select {
+	case err := <-errChan:
+		cancel()
+		testutil.SkipIfBindErr(t, err)
+		require.NoError(t, err)
+	case err := <-runChan:
+		cancel()
+		testutil.SkipIfBindErr(t, err)
+		require.NoError(t, err)
+	}
+
+	// Read generated lock and keystores from disk
+	var (
```

Comment:
@corverroos I dont see using var or const with brackets often in Go.

Reply:
It is recommended to group variables declared in multiple lines for better readability

Reply:
Where is it recommended? I have seen the opposite in Go codebases:

https://github.com/prysmaticlabs/prysm/blob/ff40a682150cbe12f044e8d636d682b68fe6bffb/beacon-chain/state/stategen/errors.go#L5-L7

https://github.com/ethereum/go-ethereum/blob/master/ethclient/ethclient.go#L112-L113

Reply:
You can find it in following:
* https://github.com/ethereum/go-ethereum/blob/master/core/beacon/errors.go
* https://github.com/prysmaticlabs/prysm/blob/ff40a682150cbe12f044e8d636d682b68fe6bffb/cmd/beacon-chain/flags/base.go#L10
* https://github.com/prysmaticlabs/prysm/blob/9615484fe9a0312d76ac68a4a4086e64f0bdfb71/io/logs/stream.go#L17
* https://github.com/prysmaticlabs/prysm/blob/9615484fe9a0312d76ac68a4a4086e64f0bdfb71/async/event/feed_test.go#L120